src/main/java/it/gov/pagopa/payment/service/qrcode/QRCodePreAuthServiceImpl.java

```diff
@@ -0,0 +1,67 @@
+package it.gov.pagopa.payment.service.qrcode;
+
+import it.gov.pagopa.payment.connector.rest.reward.RewardCalculatorConnector;
+import it.gov.pagopa.payment.connector.rest.reward.mapper.AuthPaymentMapper;
+import it.gov.pagopa.payment.dto.RewardPreview;
+import it.gov.pagopa.payment.dto.mapper.TransactionInProgress2TransactionResponseMapper;
+import it.gov.pagopa.payment.dto.qrcode.TransactionResponse;
+import it.gov.pagopa.payment.enums.SyncTrxStatus;
+import it.gov.pagopa.payment.exception.ClientExceptionWithBody;
+import it.gov.pagopa.payment.model.TransactionInProgress;
+import it.gov.pagopa.payment.repository.TransactionInProgressRepository;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+@Service
+public class QRCodePreAuthServiceImpl implements QRCodePreAuthService {
+
+  private final TransactionInProgressRepository transactionInProgressRepository;
+  private final TransactionInProgress2TransactionResponseMapper
+      transactionInProgress2TransactionResponseMapper;
+  private final AuthPaymentMapper authPaymentMapper;
+  private final RewardCalculatorConnector rewardCalculatorConnector;
+
+  public QRCodePreAuthServiceImpl(
+      TransactionInProgressRepository transactionInProgressRepository,
+      TransactionInProgress2TransactionResponseMapper
+          transactionInProgress2TransactionResponseMapper,
+      AuthPaymentMapper authPaymentMapper,
+      RewardCalculatorConnector rewardCalculatorConnector) {
+    this.transactionInProgressRepository = transactionInProgressRepository;
+    this.transactionInProgress2TransactionResponseMapper =
+        transactionInProgress2TransactionResponseMapper;
+    this.authPaymentMapper = authPaymentMapper;
+    this.rewardCalculatorConnector = rewardCalculatorConnector;
+  }
+
+  @Override
+  public TransactionResponse relateUser(String trxCode, String userId) {
+    TransactionInProgress trx = transactionInProgressRepository.findByTrxCode(trxCode);
+
+    if (trx == null) {
+      throw new ClientExceptionWithBody(
+          HttpStatus.NOT_FOUND,
+          "NOT FOUND",
+          "Cannot find transaction with trxCode [%s]".formatted(trxCode));
+    }
+
+    if (trx.getUserId() != null && !userId.equals(trx.getUserId())) {
+      throw new ClientExceptionWithBody(
+          HttpStatus.FORBIDDEN,
+          "FORBIDDEN",
+          "Transaction with trxCode [%s] is already assigned to another user".formatted(trxCode));
+    }
+
+    RewardPreview preview =
+        rewardCalculatorConnector.previewTransaction(
+            trx.getInitiativeId(), authPaymentMapper.rewardMap(trx));
+    if (preview.getStatus().equals(SyncTrxStatus.REJECTED)) {
+      transactionInProgressRepository.updateTrxRejected(
+          trx.getId(), userId, preview.getRejectionReasons());
+      throw new ClientExceptionWithBody(
```

Comment:
the transaction could be REJECTED, but due to other reasons (initiative rule not satisfied)

The "not onboarded" is recognized by the 403 error returned by reward-calculator
it could also be recognized by the initiativeRejectionReasons that will be equals to
List.of("NO_ACTIVE_INITIATIVES")

We should return 403 only when not onboarded, otherwise it will continue as usual